app/Commands/Concerns/InteractsWithIO.php

```diff
@@ -116,6 +118,23 @@ public function askToSelectSite(string $question): int
         );
     }
 
```

Comment:
This is a great idea!

Reply:
Are noodling with this a bit I've added [a commit](https://github.com/deliciousbrains/spinupwp-cli/pull/27/commits/e20bb0bc5e7c6f83f6e05d906510ff46f5ee2106) that creates a `SelectsServer` trait like you described. I'm uncertain how valuable this trait is, especially given that it depends on `App\Commands\Concerns\InteractsWithIO`. Though keeping `InteractsWithIO` tidy does have some value. 

I played around with the idea of using `addArguments()` but I could not come up with a solution that didn't smell bad. `Illuminate\Console` seems to be written in such a way as to strongly favour fluent command signatures.

Reply:
I see what you mean about it depending on `InteractsWithIO`. That in itself might point to a hurdle in future refactoring efforts, and be worth resolving. 

Would it be worth digging a little deeper into refactoring `InteractsWithIO` and moving a few of the methods out into something abstract and usable in other traits? 

I'm thinking about this using the `askToSelect()` method as an example, and think it would be fairly straightforward to move it into a helpers file by injecting the `\Illuminate\Console\OutputStyle` dependency instead of referencing it directly. 


> I played around with the idea of using addArguments() but I could not come up with a solution that didn't smell bad. Illuminate\Console seems to be written in such a way as to strongly favour fluent command signatures.

That's fair enough. Are there any specific concerns or problems that stood out to you?

Reply:
> Would it be worth digging a little deeper into refactoring InteractsWithIO and moving a few of the methods out into something abstract and usable in other traits?

Perhaps we need a rule of thumb and further discussion about what belongs in `InteractsWithIO`. There's an argument to be made that most of the parts could be refactored into individual traits but at some point we may see diminishing returns in terms of maintainability. IMHO anything that is universal should (ex. `AskToSelect`) live in that trait, whereas methods that have specific usage (eg. `askToSelectServer`, `askToSelectSite`) could have their own traits. 

Also I'm concerned about too much out of scope refactoring work here as it may impact work in progress.

> That's fair enough. Are there any specific concerns or problems that stood out to you?

Initially I thought about including the `server_id` argument in the `SelectsServer` trait because a command which selects servers will always require it. The only way I could find to accomplish this was to implement `addArgument()` in the trait's constructor. Writing a constructor into a trait is problematic because now you have a trait that's dependant on a specific parent class signature. 

As I refactored the commands to make use of this trait I realized that it would be possible to refactor the command's entire `$signature` into traits. I went down this path a little ways and found that 1) it became more difficult understand the command's params; 2) I ran into some commands where the trait didn't quite fit.

Reply:
Interesting thought experiment here, and there could be something in it, but you are right that it's probably way out of scope. Tabling this for now and we can consider revisiting down the road as the project evolves and grows.